nixos: make GIO_EXTRA_MODULES a session variable, take two

ncfavier opened the pull request:
Another attempt at #126832, which resulted in GDM crashing because of https://gitlab.gnome.org/GNOME/gdm/-/issues/756. As a workaround until that bug is fixed, wrap gnome-session-binary with an empty GIO_EXTRA_MODULES.

GDM and GNOME seem to work fine, although some more testing would be welcome.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ncfavier commented:
@jtojnar what are your thoughts on this? I know it's a hack but I don't think it's worth waiting for upstream to fix the issue

jtojnar commented:
Sounds okay to me but we will need to do more thorough testing to make sure it does not break anything else.

ncfavier commented:
I've tested basic functionality (log in, open a file browser) for the major display managers (lightdm, gdm, sddm) and desktop environments (gnome, plasma5, xfce, enlightenment, cinnamon, lxqt, mate) in a VM with `programs.dconf`, `services.gvfs` and `services.gnome.glib-networking` all enabled. Also since this is mostly a gnome thing I spent a bit more time fiddling around in gdm and gnome but didn't find any obvious issues.

Would this be enough or should we do some kind of call for testers?

ncfavier commented:
Less hacky fix following upstream discussion: allow gnome-session to find GDM's `gnome-login.session`.

@ofborg test gnome gnome-xorg

ncfavier committed:
nixos/gdm: add gdm to XDG_DATA_DIRS

This is necessary so that gnome-session can find GDM's gnome-login.session,
see https://gitlab.gnome.org/GNOME/gdm/-/issues/756

ncfavier committed:
nixos: make GIO_EXTRA_MODULES a session variable

Allow applications started by the systemd user session manager to find
their GIO_EXTRA_MODULES.

jtojnar commented:
Sounds good to me.

ncfavier commented:
Added an explanation in a comment.

ncfavier commented:
@ofborg test gnome gnome-xorg

jtojnar commented:
Looks like it is failing now: https://logs.nix.ci/?key=nixos/nixpkgs.151105&attempt_id=ac824d64-88cc-401f-9c60-8027adc68f6c

ncfavier commented:
Weird. I was only able to reproduce it once, now it's always succeeding

@ofborg test gnome

ncfavier commented:
Still [failing](https://logs.nix.ci/?key=nixos/nixpkgs.151105&attempt_id=625bd1f3-5edb-4fd5-9655-e2a18dbcfcb2), still succeeding locally on the same commit. I hate computers.

ncfavier commented:
So the error seems to be this:

```
machine # [   85.194565] .gnome-shell-wr[974]: JS ERROR: TypeError: loginItem is null
machine # _startTimedLogin/tasks<@resource:///org/gnome/shell/gdm/loginDialog.js:1072:17
machine # run@resource:///org/gnome/shell/gdm/batch.js:62:33
machine # runTask@resource:///org/gnome/shell/gdm/batch.js:134:51
machine # process@resource:///org/gnome/shell/gdm/batch.js:193:25
machine # nextTask@resource:///org/gnome/shell/gdm/batch.js:151:14
machine # process@resource:///org/gnome/shell/gdm/batch.js:204:18
machine # _start@resource:///org/gnome/shell/gdm/batch.js:159:14
machine # run@resource:///org/gnome/shell/gdm/batch.js:163:14
machine # _startTimedLogin@resource:///org/gnome/shell/gdm/loginDialog.js:1124:38
machine # _onTimedLoginRequested@resource:///org/gnome/shell/gdm/loginDialog.js:1131:14
```

Google shows no hits.

I'm going to try to run the test a couple more times, sorry for the noise
@ofborg test gnome

ncfavier commented:
Succeeds now. Looking at the [source code](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/41.2/js/gdm/loginDialog.js#L1071) this seems to be a timing issue: `loginItem` is null, probably because the `_timedLoginUserListHold` hold isn't being waited on, because the first task doesn't actually return the hold (this was changed in https://gitlab.gnome.org/GNOME/gnome-shell/-/commit/11965324938bc99e1507682b6f9db1eeaef1eeeb#16ed50f668d0ddd1464a281f6a1ddcaf482d5bd8_1052_1055 8 months ago). I will PR upstream with a fix, in the meantime maybe we should increase the autologin delay to 2 or 3 seconds.

I'm confident that the issue isn't related to this PR, though it is related to https://github.com/NixOS/nixpkgs/pull/150980 since this is specific to timed logins.

ncfavier commented:
Opened https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2116

ncfavier commented:
@ofborg test gnome gnome-xorg

ncfavier commented:
@ofborg test gnome

ncfavier committed:
nixos/tests: also set TimeoutStart in systemd.user.extraConfig

ncfavier commented:
@ofborg test gnome

jtojnar commented on nixos/modules/testing/test-instrumentation.nix:
```diff
@@ -109,6 +109,12 @@ in
     '';
+    systemd.user.extraConfig = ''
+      # Don't clobber the console with duplicate systemd messages.
+      ShowStatus=no
```
Won’t this make it harder to see the process status from logs?

ncfavier replied:
No idea, I just copied the config above for the system manager. Would things be different for the user manager?

jtojnar replied:
Sorry, did not notice it in the unexpanded diff. I would half expect the user values to be inherited :woman_shrugging:

Not sure if it should be there for system in the first place. The introducing commit https://github.com/ncfavier/nixpkgs/commit/eab25b104ad12b75a8b9a41ce3ac8b3b931144ef does not have much context.

ncfavier replied:
They're not inherited, I checked in a VM and the TimeoutStart was 90 seconds, hence the failure on aarch64.

Also it seems like ShowStatus actually only applies to the system manager.

ncfavier commented:
Let's see how the logs compare
@ofborg test gnome

ncfavier commented:
Yeah this is just duplicated noise
```
machine # [    5.492191] systemd[1]: Starting Journal Service...
machine #          Starting Journal Service...
```

ncfavier commented:
Should be all good now
@ofborg test gnome
@ofborg test gnome-xorg

ncfavier commented:
Now that https://gitlab.gnome.org/GNOME/gdm/-/merge_requests/168 is merged, do you want to wait for the next GDM release so we can skip d25ffc3a0c299f9ef98578898709ff7ffc031137 ?

jtojnar commented:
Not sure if it makes sense to wait https://wiki.gnome.org/FortyTwo

ncfavier commented:
Let's not wait then.

ncfavier commented:
What are we waiting for?
